### PR TITLE
feat: update Lambda's SQS event batch size and add support to batching window

### PR DESCRIPTION
**Pull request by xiqi**
Update SQS event batch size and add support to batching window according to: https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 1]
```diff
@@ -45,10 +45,13 @@ class AwsCompileSQSEvents {
             let EventSourceArn;
             let BatchSize = 10;
             let Enabled = true;
+            let MaximumBatchingWindowInSeconds = 0;
 
             if (typeof event.sqs === 'object') {
               EventSourceArn = event.sqs.arn;
               BatchSize = event.sqs.batchSize || BatchSize;
+              MaximumBatchingWindowInSeconds =
+                event.sqs.maximumBatchingWindowInSeconds || MaximumBatchingWindowInSeconds;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:54]: We need to define this option in config schema

File: docs/providers/aws/events/sqs.md [hunk 2]
```diff
@@ -49,7 +49,7 @@ functions:
 
 ## Setting the BatchSize
 
-For the SQS event integration, you can set the `batchSize`, which effects how many SQS messages can be included in a single Lambda invocation. The default `batchSize` is 10, and the max `batchSize` is 10.
+For the SQS event integration, you can set the `batchSize`, which effects how many SQS messages can be included in a single Lambda invocation. The default `batchSize` is 10, and the max `batchSize` is 10000 for a standard queue, 10 for a FIFO queue. You can also set `maximumBatchingWindowInSeconds` to standard queues to specify the maximum amount of time in seconds to gather records before invoking the function. Read more [here](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html).
```
Comment (medikoo) [inline docs/providers/aws/events/sqs.md:52]: Let's replace "Read more [here]" with "Check [AWS documentation] for more details"

**Review by medikoo — CHANGES_REQUESTED** [review]
@xiqi Thank you, that's nice improvement! Please see my remarks

File: lib/plugins/aws/package/compile/events/sqs/index.js (hunk 1 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.js:55]: Thank you! Fixed by https://github.com/serverless/serverless/commit/6c54c629bfa2638564a164c23ec35ce80ebf43bc

File: docs/providers/aws/events/sqs.md (hunk 2 above)
Comment (xiqi) [reply docs/providers/aws/events/sqs.md]: Updated: https://github.com/serverless/serverless/commit/d4f172686d62eb3e3fb8c873bcf1bdb32113eda9

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 3]
```diff
@@ -19,8 +19,9 @@ class AwsCompileSQSEvents {
           type: 'object',
           properties: {
             arn: { $ref: '#/definitions/awsArn' },
-            batchSize: { type: 'integer', minimum: 1, maximum: 10 },
+            batchSize: { type: 'integer', minimum: 1, maximum: 10000 },
             enabled: { type: 'boolean' },
+            maximumBatchingWindowInSeconds: { type: 'integer', minimum: 0 },
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:24]: Sorry, for second call. I just realized we should actually call it `maximumBatchingWindow`, to be on pair with spec proposed for new events here https://github.com/serverless/serverless/issues/8137

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 4]
```diff
@@ -45,10 +46,13 @@ class AwsCompileSQSEvents {
             let EventSourceArn;
             let BatchSize = 10;
             let Enabled = true;
+            let MaximumBatchingWindowInSeconds = 0;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:49]: Let's not set the default. If `event.sqs.maximumBatchingWindowInSeconds` is not set, let's not set property at all.

**Review by medikoo — CHANGES_REQUESTED** [review]
@xiqi great thanks for update, and sorry for another turnaround (I didn't look close enough at first time).

To be consistent we need to tackle two more things.

File: lib/plugins/aws/package/compile/events/sqs/index.js (hunk 3 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.js]: No worries!
I've changed it: https://github.com/serverless/serverless/commit/d70c58b49c8a30ec194102801727f1b570db37d0

File: lib/plugins/aws/package/compile/events/sqs/index.js (hunk 4 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.js]: Not sure whether this is the best way to do it: https://github.com/serverless/serverless/commit/d70c58b49c8a30ec194102801727f1b570db37d0#diff-bb7849521252332fd089024a0031c78cbc85ac2eda8c9d4df916582f865f9a76R99-R105

Let me know if you have a more elegant way.

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 5]
```diff
@@ -95,11 +96,18 @@ class AwsCompileSQSEvents {
               }
             `;
 
+            const sqsTemplateObject = JSON.parse(sqsTemplate);
+
+            if (typeof event.sqs.maximumBatchingWindow !== 'undefined') {
+              sqsTemplateObject.Properties.MaximumBatchingWindowInSeconds =
+                event.sqs.maximumBatchingWindow;
+            }
+
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:105]: Let's maybe convert above notation so we directly create `sqsTemplateObject` (creating JSON string, is some old controversial convention used sometimes in a Framework, and in general we're not practising it new code).

Having that configuration would be  simply as;

```javascript
const sqsTemplate = {
  Type: 'AWS::Lambda::EventSourceMapping',
  ...
  Properties: {
    BatchSize,
    ...
  }
}
```

Having that we can include `MaximumBatchingWindowInSeconds` in above construct with a form:

```javascript
  MaximumBatchingWindowInSeconds: event.sqs.maximumBatchingWindow != null ? event.sqs.maximumBatchingWindow : undefined
```

1. `!= null`  checks wether value is neither `null` nor `undefined` (it's the values for which we should assume that property is not set)
2. Assigning `undefined` is like omitting property, as `undefined` properties do not get serialized to JSON.

File: docs/providers/aws/events/sqs.md [hunk 6]
```diff
@@ -49,7 +49,7 @@ functions:
 
 ## Setting the BatchSize
 
-For the SQS event integration, you can set the `batchSize`, which effects how many SQS messages can be included in a single Lambda invocation. The default `batchSize` is 10, and the max `batchSize` is 10.
+For the SQS event integration, you can set the `batchSize`, which effects how many SQS messages can be included in a single Lambda invocation. The default `batchSize` is `10`. The max `batchSize` is `10000` for a standard queue, `10` for a FIFO queue. You can also set `maximumBatchingWindow` to standard queues to specify the maximum amount of time in seconds to gather records before invoking the function. Check [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html) for more details.
```
Comment (medikoo) [inline docs/providers/aws/events/sqs.md:52]: Let's put `You can also set..` in new line (it'll give more breath)

**Review by medikoo — COMMENTED** [review]
Thank you @xiqi looks really good. I've proposed just some cleanup, so we land with more straightforward construct.

Let me know what you think

File: lib/plugins/aws/package/compile/events/sqs/index.js (hunk 5 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.js]: Thank you so much for the help! I've updated it: https://github.com/serverless/serverless/commit/c9da466875da95584c754b991ec1646998f10018

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 7]
```diff
@@ -19,8 +19,9 @@ class AwsCompileSQSEvents {
           type: 'object',
           properties: {
             arn: { $ref: '#/definitions/awsArn' },
-            batchSize: { type: 'integer', minimum: 1, maximum: 10 },
+            batchSize: { type: 'integer', minimum: 1, maximum: 10000 },
             enabled: { type: 'boolean' },
+            maximumBatchingWindow: { type: 'integer', minimum: 0 },
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:24]: Let's also configure `maximum` (If I see correctly it's `300` per docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html#cfn-lambda-eventsourcemapping-maximumbatchingwindowinseconds)

**Review by medikoo — COMMENTED** [review]
@xiqi it looks great! I've just spotted one omission in schema configuration, and can we have new property be backed by unit test? For that ideally extend already existing `runServerless` run here: https://github.com/serverless/serverless/blob/c9da466875da95584c754b991ec1646998f10018/lib/plugins/aws/package/compile/events/sqs/index.test.js#L594-L605

1. Configure new function (in `configExt` property), which would have `sns` event and `maximumBatchingWindow` property
2. Then in dedicated `it` confirm that generated CF resource has expected property and value

For more info on tests, check: https://github.com/serverless/serverless/tree/c9da466875da95584c754b991ec1646998f10018/test#unit-tests

File: lib/plugins/aws/package/compile/events/sqs/index.js (hunk 7 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.js]: Good catch! I've added the maximum value check as well as the unit test

File: lib/plugins/aws/package/compile/events/sqs/index.test.js [hunk 8]
```diff
@@ -618,5 +626,15 @@ describe('AwsCompileSQSEvents #2', () => {
       const aliasLogicalId = naming.getLambdaProvisionedConcurrencyAliasLogicalId('foo');
       expect(eventSourceMappingResource.DependsOn).to.include(aliasLogicalId);
     });
+
+    it('should have correct batch size', () => {
+      expect(JSON.stringify(eventSourceMappingResource.Properties.BatchSize)).to.equal('10');
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.test.js:631]: Let's not use `JSON.stringify`. Simply confirm literal values

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @xiqi looks great! I've just proposed to not use `JSON.stringify` it doesn't seem justified to confirm on JSON stringified values

File: lib/plugins/aws/package/compile/events/sqs/index.test.js (hunk 8 above)
Comment (xiqi) [reply lib/plugins/aws/package/compile/events/sqs/index.test.js]: Agreed and updated

**Review by medikoo — APPROVED** [review]
Thank you @xiqi !

**Pull request merged**